codec/enforceproto/extraneous_fields_test.go

```diff
@@ -0,0 +1,413 @@
+	tests := []struct {
+		name    string
+		in      proto.Message
+		recv    proto.Message
+		wantErr error
+	}{
+		{
+			name: "TestVersion3 from TestVersionFD1",
+			in: &testdata.TestVersion2{
+				X: 5,
+				Sum: &testdata.TestVersion2_E{
+					E: 100,
+				},
+				H: []*testdata.TestVersion1{
+					{X: 999},
+					{X: -55},
+					{
+						X: 102,
+						Sum: &testdata.TestVersion1_F{
+							F: &testdata.TestVersion1{
+								X: 4,
+							},
+						},
+					},
+				},
+				Customer1: &testdata.Customer1{
+					Id:              45,
+					Name:            "customer1",
+					SubscriptionFee: 99,
+				},
+			},
+			recv: new(testdata.TestVersionFD1),
+			wantErr: &errExtraneousField{
+				Type:     "*testdata.TestVersionFD1",
+				TagNum:   12,
+				WireType: 2,
+			},
+		},
+		{
+			name: "Nested2B vs Nested3B",
+			in: &testdata.Nested2B{
+				Id:  5,
+				Fee: 88.91,
+				Nested: &testdata.Nested3B{
+					Id:   15,
+					Age:  47,
+					Name: "Crowley",
+					B4: []*testdata.Nested4B{
+						{
+							Id: 34, Age: 55, Name: "Trivia",
+						},
+						{
+							Id: 55, Age: 78, Name: "Incredible",
+						},
+					},
+				},
+			},
+			recv: new(testdata.TestVersion3),
+			wantErr: &errMismatchedWireType{
+				Type:         "*testdata.TestVersion3",
+				TagNum:       2,
+				GotWireType:  1,
+				WantWireType: 2,
+			},
+		},
+		{
+			name: "Alternating oneofs",
+			in: &testdata.TestVersion3{
+				Sum: &testdata.TestVersion3_E{
+					E: 99,
+				},
+			},
+			recv:    new(testdata.TestVersion3LoneOneOfValue),
+			wantErr: nil,
+		},
+		{
+			name: "Alternating oneofs mismatched field",
+			in: &testdata.TestVersion3{
+				Sum: &testdata.TestVersion3_F{
+					F: &testdata.TestVersion3{
+						X: 99,
+					},
+				},
+			},
+			recv: new(testdata.TestVersion3LoneOneOfValue),
+			wantErr: &errExtraneousField{
+				Type:     "*testdata.TestVersion3LoneOneOfValue",
+				TagNum:   7,
+				WireType: 2,
+			},
+		},
+		{
+			name: "Discrepancy in a deeply nested one of field",
+			in: &testdata.TestVersion3{
+				Sum: &testdata.TestVersion3_F{
+					F: &testdata.TestVersion3{
+						Sum: &testdata.TestVersion3_F{
+							F: &testdata.TestVersion3{
+								X: 19,
+								Sum: &testdata.TestVersion3_E{
+									E: 99,
+								},
+							},
+						},
+					},
+				},
+			},
+			recv: new(testdata.TestVersion3LoneNesting),
+			wantErr: &errExtraneousField{
+				Type:     "*testdata.TestVersion3LoneNesting",
+				TagNum:   6,
+				WireType: 0,
+			},
+		},
+		{
+			name: "types.Any in G",
+			in: &types.Any{
+				TypeUrl: "/testdata.TestVersion1",
+				Value: mustMarshal(&testdata.TestVersion3{
+					X: 102,
+					Sum: &testdata.TestVersion3_F{
+						F: &testdata.TestVersion3{
+							X: 4,
+						},
+					},
+				}),
+			},
+			recv: new(testdata.TestVersion3LoneNesting),
+			wantErr: &errMismatchedWireType{
+				Type:         "*testdata.TestVersion3LoneNesting",
+				TagNum:       1,
+				GotWireType:  2,
+				WantWireType: 0,
```

Comment:
Let's actually test a failure case that has an unexpected field. As I've said before, I am unconcerned about mismatched wire types. I want to know if there are unexpected fields. So `recv` should be `Any` and we could pass `TestVersion2` in with `new_field` set but the `TypeURL` for `TestVersion1`. That would actually check an `Any` failure case.

Reply:
Thank you @aaronc, I've addressed this feedback and added the appropriate tests.